src/dscanner/utils.d

```diff
@@ -0,0 +1,194 @@
+	ubyte[4096] buf;
+	while (true)
+	{
+		auto b = stdin.rawRead(buf);
+		if (b.length == 0)
+			break;
+		sourceCode.put(b);
+	}
+	return sourceCode.data;
+}
+
+ubyte[] readFile(string fileName)
+{
+	if (fileName == "stdin")
+		return readStdin();
+	if (!exists(fileName))
+	{
+		stderr.writefln("%s does not exist", fileName);
+		return [];
+	}
+	File f = File(fileName);
+	if (f.size == 0)
+		return [];
+	ubyte[] sourceCode = uninitializedArray!(ubyte[])(to!size_t(f.size));
+	f.rawRead(sourceCode);
+	return sourceCode;
+}
+
+string[] expandArgs(string[] args)
+{
+	import std.file : isFile, FileException, dirEntries, SpanMode;
+	import std.algorithm.iteration : map;
+	import std.algorithm.searching : endsWith;
+
+	// isFile can throw if it's a broken symlink.
+	bool isFileSafe(T)(T a)
+	{
+		try
+			return isFile(a);
+		catch (FileException)
+			return false;
+	}
+
+	string[] rVal;
+	if (args.length == 1)
+		args ~= ".";
+	foreach (arg; args[1 .. $])
+	{
+		if (arg == "stdin" || isFileSafe(arg))
+			rVal ~= arg;
+		else
+			foreach (item; dirEntries(arg, SpanMode.breadth).map!(a => a.name))
+			{
+				if (isFileSafe(item) && (item.endsWith(`.d`) || item.endsWith(`.di`)))
+					rVal ~= item;
+				else
+					continue;
+			}
+	}
+	return rVal;
+}
+
+/**
+ * Allows to build access chains of class members as done with the $(D ?.) operator
+ * in other languages. In the chain, any $(D null) member that is a class instance
+ * or that returns one, has for effect to shortcut the complete evaluation.
+ *
+ * This function is copied from https://github.com/BBasile/iz to avoid a new submodule.
+ * Any change made to this copy should also be applied to the origin.
+ *
+ * Params:
+ *      M = The class type of the chain entry point.
+ *
+ * Bugs:
+ *      Assigning a member only works with $(D unwrap).
+ *
+ */
+struct SafeAccess(M)
+if (is(M == class))
+{
+	M m;
+
+	@disable this();
+
+	/**
+	 * Instantiate.
+	 *
+	 * Params:
+	 *      m = An instance of the entry point type. It is usually only
+	 *      $(D null) when the constructor is used internally, to build
+	 *      the chain.
+	 */
+	this(M m)
+	{
+		this.m = m;
+	}
+
+	alias m this;
+	/// Unprotect the class instance.
+	alias unwrap = m;
+
+	/// Handles safe access.
+	auto ref opDispatch(string member, A...)(auto ref A a)
+	{
+		import std.traits : ReturnType;
+		alias T = typeof(__traits(getMember, m, member));
+		static if (is(T == class))
+		{
+			return (!m || !__traits(getMember, m, member))
+				? SafeAccess!T(null)
+				: SafeAccess!T(__traits(getMember, m, member));
+		}
+		else
+		{
+			import std.traits : ReturnType, Parameters, isFunction;
+			static if (isFunction!T)
+			{
+				// otherwise there's a missing return statement.
+				alias R = ReturnType!T;
+				static if (!is(R == void) &&
+					!(is(R == class) && Parameters!T.length == 0))
+						pragma(msg, __FILE__ ~ "(" ~ __LINE__.stringof ~ "): error, " ~
+						"only `void function`s or `class` getters can be called without unwrap");
```

Comment:
shouldn't this be an `static assert(0`, s.t. the compilation stops?
+ nit: use backticks with `unwrap`, s.t. DMD's colored highlighting works

Reply:
Also why can't we return the result here?

Reply:
> shouldn't this be an static assert(0, s.t. the compilation stops?

the problem is that opDispatch gags the error messages so what happens exactly is that i let the error happens and use pragma to emit a pseudo error message.